### PR TITLE
Speed up PaddedBytes with block writes instead of per-byte loop

### PR DESCRIPTION
**Pull request by kamilchodola**
## Summary

Speeds up `UInt256.PaddedBytes(int)` by replacing its per-byte loop with block big-endian writes, while preserving its truncation and zero-padding behavior.

## Change

- Writes the 20- and 32-byte forms directly through the existing `ToBigEndian(Span<byte>)` paths.
- For wider fresh arrays, writes only the right-aligned 32-byte tail because the leading bytes are already zero.
- Uses a fully initialized 32-byte stack buffer for other widths, then copies the required low bytes.
- Adds `WritePaddedBytes(Span<byte>)` for allocation-free caller-owned buffers. The distinct verb avoids overload ambiguity with `PaddedBytes(int)` for calls such as `PaddedBytes(default)`.

## Benchmark (`PaddedBytesUnsigned`, .NET 10, `[MemoryDiagnoser]`)

The benchmark is an in-process A/B against the previous per-byte implementation and lets BenchmarkDotNet calibrate the invocation count.

| n | Previous | Block writes | No-intrinsics previous | No-intrinsics block writes |
|---|---:|---:|---:|---:|
| 20 (address) | 42.54 ns | 4.58 ns | 45.43 ns | 4.63 ns |
| 32 (EVM word) | 67.19 ns | 4.38 ns | 70.56 ns | 5.06 ns |
| 33 | 58.70 ns | 4.90 ns | 71.92 ns | 5.58 ns |
| 64 | 70.70 ns | 5.24 ns | 78.66 ns | 6.42 ns |

Allocation is unchanged for `PaddedBytes(int)`, which still returns a fresh array by contract. `WritePaddedBytes(Span<byte>)` allocates 0 B.

## Validation

- Full merged test suite passes: 575,765 tests.
- Five million random values across widths 0–40 produced byte-identical output versus the previous implementation, plus explicit boundary and limb-pattern cases.
- Regression tests cover right alignment, truncation, wide-span zeroing, dirty targets, and explicit byte-for-byte output for the delegated 20- and 32-byte paths.
- Library builds in Release with 0 warnings; benchmark builds with only pre-existing warnings.

### CODE REVIEW

**Comment by smartprogrammer93** [comment]
Claude review:

Verdict: Approve with nits. Functionally equivalent to the per-byte loop for all n — traced the old loop's limb-by-limb fill against the new big-endian block writes and limb order (u3, u2, u1, u0) matches the existing ToBigEndian. Right-alignment, truncation when n < 32, and zero-padding when n > 32 are all preserved. Solid ~3× win.

Findings
- Should-fix: The new PaddedBytes(Span<byte>) overload at UInt256.Conversions.cs:51-67 leaves the leading n - 32 bytes of the caller's buffer untouched when n > 32. The <summary> mentions it, but there's no <param>, no <remarks>, no <exception> — and this is the kind of asymmetry vs. the byte[] overload that will bite a caller passing a re-used 64-byte buffer. Recommend either (a) zeroing target.Slice(0, n - copy) for behavioral parity with the array overload (essentially free; the array overload still wins because the JIT can elide the redundant zero on a fresh array), or (b) renaming to WritePaddedBytes and documenting the divergence loudly. Lean (a).
- Should-fix: PaddedBytes(int) still has no XML doc. Since the new overload sits beside it with docs, either add <summary>/<param>/<returns> to both, or <inheritdoc cref="PaddedBytes(Span{byte})"/> on the array overload.
- Nits:
  - Tests at UInt256Tests.cs:651-675 cover good n boundaries (0, 1, 8, 20, 31, 32, 33, 64) but use a single fixed mid-sized value. The PR body claims "max, every boundary width" coverage — actually exercising UInt256.MaxValue and UInt256.Zero at each width via [TestCaseSource] would lock down the all-ones top-limb and all-zeros paths.
  - The Span overload test (UInt256Tests.cs:678-694) only exercises n > 32. Add an n < 32 case writing into a pre-filled buffer to pin the right-alignment + tail-overwrite semantics.

Question for the author: Re open-question #2 — I'd lean toward zeroing the leading bytes in the span overload. The "caller owns zeroing" contract saves nothing measurable at these sizes and creates a surprising asymmetry with the array overload. Is there a specific callsite that motivated the no-zero behavior?

**Comment by kamilchodola** [comment]
Thanks for the review! Addressed in 913fd64:

- **Should-fix (span overload leading bytes):** went with option (a) — `PaddedBytes(Span<byte>)` now zeroes the leading `n-32` bytes when `n > 32`, giving full-span behavioral parity with the `byte[]` overload. No more asymmetry for a caller passing a reused buffer. (The array overload starts from a zeroed array so the extra `Clear` is elided there.) No specific callsite motivated the no-zero behavior — it was just an under-specified initial contract, so parity is the right call.
- **Should-fix (XML doc on `PaddedBytes(int)`):** added `<summary>`/`<param>`/`<returns>` to both overloads.
- **Nit (single fixed value):** the right-aligned matrix now runs across `Zero`, `MaxValue` (all-ones top limb), and the mid-sized value × all boundary widths via `[ValueSource]`/`[Values]`.
- **Nit (span test only n > 32):** added an `n < 32` case writing into a pre-filled buffer to pin the truncation + right-alignment + full-overwrite semantics, alongside the existing `n > 32` leading-zero case.

**Comment by kamilchodola** [comment]
## EVM-level gas benchmark — context for this PR

Ran an int256-focused gas-benchmark of the bundled #84/#85/#86 changes vs `master` (5 runs each, 18 EVM fixtures; runs master `26818163268` · branch `26817551249`). Overall **−0.2% (148.9 → 148.5 ms)** — performance-neutral at the block layer, with the only beyond-noise wins being mod-arithmetic (`MULMOD mod_bits_191 −9.3%`, `MOD mod_bits_255 −2.9%`), which belong to #84.

**Scoping note for this PR specifically:** none of the 18 EVM fixtures exercise `PaddedBytes`, so this gas-benchmark says nothing direct about this change — it serves only as a **safety check** that, bundled with the other two, this PR causes no EVM-layer regression (confirmed: no regression attributable here).

The real evidence for this PR is the `PaddedBytes` microbenchmark in the PR body:

> `PaddedBytesUnsigned` — **~2–3× faster** (~296–362 ns → ~97–177 ns) across n ∈ {20, 32, 33, 64}, same speedup in the no-intrinsics job.

That win is at the library level (replacing the per-byte limb-indexer loop with block writes) and was validated byte-for-byte against the previous implementation with a 5M-case fuzz plus the unit-test matrix on this branch. It's simply below the granularity a block-level gas benchmark resolves.

File: src/Nethermind.Int256/UInt256.Conversions.cs [hunk 1]
```diff
@@ -39,16 +39,49 @@ public object ToType(Type conversionType, IFormatProvider? provider) =>
...
 
-        for (int i = 0; i < 32 && i < n; i++)
+    /// <summary>
+    /// Writes the big-endian, right-aligned, left-zero-padded representation of this value into
+    /// <paramref name="target"/>, filling the entire span. When the target is shorter than 32 bytes
+    /// the most-significant bytes are truncated; when it is longer the extra leading bytes are
+    /// zeroed, matching the <see cref="PaddedBytes(int)"/> array overload.
+    /// </summary>
+    /// <param name="target">The destination span; fully overwritten.</param>
+    public void PaddedBytes(Span<byte> target)
```
Comment (LukaszRozmej) [inline src/Nethermind.Int256/UInt256.Conversions.cs]:
```suggestion
    [SkipLocalsInit]
    public void PaddedBytes(Span<byte> target)
```

**Review by benaadams — APPROVED** [review]
None of these paths are used in any fast path

**Review by LukaszRozmej — COMMENTED** [review]
Block-write rewrite looks correct.

A few notes building on the `[SkipLocalsInit]` suggestion already inline:

1. **PR description is out of sync with the implementation.** The "Open questions" point #2 says the span overload "does not zero any extra leading bytes when n > 32", but the implementation does zero them (and `PaddedBytes_Into_Span_ZeroesLeadingBytes_When_Longer_Than_Word` asserts that). The current behavior matches the array overload — keep it, just fix the description.

2. **Minor:** the comment "the JIT elides this there" on the byte[] path is overstated. `Span<byte>.Clear()` isn't visible to the JIT as a no-op even when the underlying array is freshly zero-initialized — the JIT can't prove that across the method boundary. Not a real bottleneck (n > 32 is uncommon), but consider either splitting the byte[] overload to skip the redundant Clear, or toning down the comment.

3. **Test coverage** is solid — the `[ValueSource]` × `[Values]` cross-product pins the boundary cases well.

**Review by LukaszRozmej — APPROVED** [review]

**Comment by kamilchodola** [comment]
Addressed in c0051fc:
- Applied the `[SkipLocalsInit]` suggestion to `PaddedBytes(Span<byte>)`.
- Trimmed the two inline comments — the block-write rationale duplicated the XML doc, and the "JIT elides this" claim was overstated per your note, so I just dropped both rather than rewording.
- Fixed the PR description: open question #2 was stale (said the span overload doesn't zero leading bytes; it does, matching the array overload) — updated to describe actual behavior.

**Comment by kamilchodola** [comment]
**zkVM follow-up** (verified by an adversarial multi-agent review of the guest-executed path): the staging-buffer shape was 3-4x faster than master for the dominant widths but had two guest-visible (and host-visible) inefficiencies, fixed in 3c39c0e:
- `PaddedBytes(Span)` now delegates `n == 32` and `n == 20` to the existing `ToBigEndian(Span)`, which writes the reversed limbs straight into the target — removes the 32-byte stackalloc staging buffer and the `CopyTo` (~8 loads + 8 stores + memmove dispatch per call) for the two widths that dominate real usage (EVM word, address). Bonus: if #92 lands, these widths automatically pick up its vectorized write on AVX2 hosts.
- `PaddedBytes(int)` with `n > 32` now writes only the right-aligned 32-byte tail of the fresh array, so the span overload's leading-bytes `Clear` no longer re-zeroes memory `new byte[n]` already guarantees is zero.

Remaining known edge: for tiny widths (n ≤ ~8, no known callers) the block-write path still reverses all four limbs before copying n bytes — on a no-Zbb RISC-V zk guest that's ~5x master at n=1. Left as-is for simplicity; gating each limb write on the copy window would fix it if such callers ever appear. Full 575k-test suite green.

**Review by benaadams — APPROVED** [review]
